pkg/relay/ethereum/tx.go

```diff
@@ -460,3 +429,265 @@ func (c *Chain) parseRpcError(err error) (string, string) {
 	// Note that Raw error data may be available even if revert reason isn't available.
 	return revertReason, hex.EncodeToString(rawErrorData)
 }
+
+
+func estimateGas(
+	ctx context.Context,
+	c *Chain,
+	tx *gethtypes.Transaction,
+	doRound bool, // return rounded gas limit when gas limit is over
+	base_logger *log.RelayLogger,
+) (uint64, error) {
+	logger := &log.RelayLogger{ Logger: base_logger.Logger }
+
+	if rawTxData, err := tx.MarshalBinary(); err != nil {
+		logger.Error("failed to encode tx", err)
+	} else {
+		logger.Logger = logger.With(logAttrRawTxData, hex.EncodeToString(rawTxData))
+	}
+
+	estimatedGas, err := c.client.EstimateGasFromTx(ctx, tx)
+	if err != nil {
+		if revertReason, rawErrorData, err := c.getRevertReasonFromRpcError(err); err != nil {
+			// Raw error data may be available even if revert reason isn't available.
+			logger.Logger = logger.With(logAttrRawErrorData, hex.EncodeToString(rawErrorData))
+			logger.Error("failed to get revert reason", err)
+		} else {
+			logger.Logger = logger.With(
+				logAttrRawErrorData, hex.EncodeToString(rawErrorData),
+				logAttrRevertReason, revertReason,
+			)
+		}
+
+		logger.Error("failed to estimate gas", err)
+		return 0, err
+	}
+
+	txGasLimit := estimatedGas * c.Config().GasEstimateRate.Numerator / c.Config().GasEstimateRate.Denominator
+	if txGasLimit > c.Config().MaxGasLimit {
+		if !doRound {
+			return 0, fmt.Errorf("estimated gas exceeds max gas limit")
+		}
+
+		logger.Warn("estimated gas exceeds max gas limit",
+			logAttrEstimatedGas, txGasLimit,
+			logAttrMaxGasLimit, c.Config().MaxGasLimit,
+		)
+		return c.Config().MaxGasLimit, nil
+	}
+
+	return txGasLimit, nil
+}
+
+
+type CallIter struct {
+	msgs []sdk.Msg
+	cursor int
+	skipUpdateClientCommitment bool
+	// for multicall
+	txs []gethtypes.Transaction
+	msgTypeNames []string
+}
+func NewCallIter(msgs []sdk.Msg, skipUpdateClientCommitment bool) CallIter {
+	return CallIter {
+		msgs: msgs,
+		cursor: 0,
+		skipUpdateClientCommitment: skipUpdateClientCommitment,
+	}
+}
+func (iter *CallIter) Cursor() int {
+	return iter.cursor
+}
+func (iter *CallIter) Current() *sdk.Msg {
+	return &iter.msgs[iter.cursor]
+}
```

Comment:
`sdk.Msg` is not struct but interface.
We need not use a pointer to the interface here.

Reply:
fixed